lib/Dialect/FIRRTL/Transforms/InferResets.cpp

```diff
@@ -0,0 +1,1538 @@
+          (drive.src->type.isa<AsyncResetType>() && !majorityAsync) ||
+          (drive.dst->type.isa<UIntType>() && majorityAsync) ||
+          (drive.src->type.isa<UIntType>() && majorityAsync))
+        diag.attachNote(drive.loc)
+            << "Offending " << (majorityAsync ? "sync" : "async")
+            << " drive here:";
+    }
+    return failure();
+  }
+
+  // At this point we know that the type of the reset is unambiguous. If there
+  // are any votes for async, we make the reset async. Otherwise we make it
+  // sync.
+  net->kind = (asyncDrives ? ResetKind::Async : ResetKind::Sync);
+  LLVM_DEBUG(llvm::dbgs() << "- Inferred as " << net->kind << "\n");
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// Reset Updating
+//===----------------------------------------------------------------------===//
+
+LogicalResult InferResetsPass::updateResets() {
+  for (auto *net : resetMap.getNets()) {
+    if (failed(updateReset(net)))
+      return failure();
+  }
+  return success();
+}
+
+LogicalResult InferResetsPass::updateReset(ResetNet *net) {
+  LLVM_DEBUG(llvm::dbgs() << "Updating reset network with " << net->nodes.size()
+                          << " nodes to " << net->kind << "\n");
+  assert(net->kind != ResetKind::Uninferred &&
+         "all reset nets should be inferred at this point");
+
+  // Determine the final type the reset should have.
+  FIRRTLType resetType;
+  if (net->kind == ResetKind::Async)
+    resetType = AsyncResetType::get(&getContext());
+  else
+    resetType = UIntType::get(&getContext(), 1);
+
+  // Update all those values in the network that cannot be inferred from
+  // operands. If we change the type of a module port (i.e. BlockArgument), add
+  // the module to a module worklist since we need to update its function type.
+  SmallSetVector<Operation *, 16> worklist;
+  SmallDenseSet<Operation *> moduleWorklist;
+  for (auto node : net->nodes) {
+    Value value = node->value.getValue();
+    if (!value.isa<BlockArgument>() &&
+        !isa_and_nonnull<WireOp, RegOp, RegResetOp, InstanceOp, InvalidValueOp>(
+            value.getDefiningOp()))
+      continue;
+    if (updateReset(node->value, resetType)) {
+      for (auto user : value.getUsers())
+        worklist.insert(user);
+      if (auto blockArg = value.dyn_cast<BlockArgument>())
+        moduleWorklist.insert(blockArg.getOwner()->getParentOp());
+    }
+  }
+
+  // Work dat list.
+  while (!worklist.empty()) {
+    auto op = dyn_cast_or_null<InferTypeOpInterface>(worklist.pop_back_val());
+    if (!op)
+      continue;
+
+    // Determine the new result types.
+    SmallVector<Type, 2> types;
+    if (failed(op.inferReturnTypes(op->getContext(), op->getLoc(),
+                                   op->getOperands(), op->getAttrDictionary(),
+                                   op->getRegions(), types)))
+      return failure();
```

Comment:
How and when do the return types of operations change?

Reply:
Mainly `SubfieldOp`, `SubindexOp`, and `SubaccessOp`, for example when a `!firrtl.reset` type is buried somewhere in a bundle port. Note that this runs only on the fanout cone of the stuff that contains a reset, so should only call `inferReturnTypes` on a very small number of operations in the IR.

Reply:
Got it, thanks!